File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/DiaryYearMonthListBaseItem.java
package com.websarva.wings.android.zuboradiary.ui.list;

import androidx.annotation.NonNull;

import com.websarva.wings.android.zuboradiary.ui.list.DiaryYearMonthListBaseAdapter.ViewType;

import java.time.YearMonth;
import java.util.Objects;

public abstract class DiaryYearMonthListBaseItem {
    private final YearMonth yearMonth;
    private final ViewType viewType;

    public DiaryYearMonthListBaseItem(ViewType viewType) {
        Objects.requireNonNull(viewType);

        yearMonth = YearMonth.now();
        this.viewType = viewType;
    }

    public DiaryYearMonthListBaseItem(YearMonth yearMonth, ViewType viewType) {
        Objects.requireNonNull(yearMonth);
        Objects.requireNonNull(viewType);

        this.yearMonth = yearMonth;
        this.viewType = viewType;
    }

    public final boolean isNotDiaryViewType() {
        return !viewType.equals(ViewType.DIARY);
    }

    @NonNull
    public final YearMonth getYearMonth() {
        return this.yearMonth;
    }

    public final ViewType getViewType() {
        return viewType;
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/adapter/diary/DiaryYearMonthListBaseItem.kt
package com.websarva.wings.android.zuboradiary.ui.adapter.diary

import java.time.YearMonth

internal abstract class DiaryYearMonthListBaseItem {

    val yearMonth: YearMonth
    val viewType: DiaryYearMonthListBaseAdapter.ViewType
    abstract val diaryDayList: DiaryDayBaseList

    val isNotDiaryViewType
        get() = viewType != DiaryYearMonthListBaseAdapter.ViewType.DIARY

    constructor(viewType: DiaryYearMonthListBaseAdapter.ViewType): this(YearMonth.now(), viewType)

    constructor(yearMonth: YearMonth, viewType: DiaryYearMonthListBaseAdapter.ViewType) {
        this.yearMonth = yearMonth
        this.viewType = viewType
    }


    fun areItemsTheSame(item: DiaryYearMonthListBaseItem): Boolean {
        if (this === item) return true

        return yearMonth == item.yearMonth && viewType == item.viewType
    }

    abstract fun areContentsTheSame(item: DiaryYearMonthListBaseItem): Boolean

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is  DiaryYearMonthListBaseItem) return false

        return yearMonth == other.yearMonth && viewType == other.viewType
    }

    override fun hashCode(): Int {
        var result = yearMonth.hashCode()
        result = 31 * result + viewType.hashCode()
        return result
    }
}
